### DISCUSSION TITLE
CC: Tweaked 1.100 for MC 1.16.x, 1.17.x and 1.18.x

### DISCUSSION THREAD

**Discussion by SquidDev** [cc-tweaked/CC-Tweaked#998] [Announcements]
Merry Christmas!

If I've got my maths right, and it's going to be terribly embarrassing if I haven't, ComputerCraft is 10 years old today. Not only is this an incredible achievement in its own right, [tradition mandates](https://github.com/cc-tweaked/CC-Tweaked/discussions/625)[^1] that it's time for another CC: Tweaked release!

CC: Tweaked 1.100's[^2] changelog is pretty small, but does contain one new feature: Speakers!

## Speakers!

Speakers were one of the first new features added to ComputerCraft after it being open sourced, [contributed by Restioson](https://github.com/dan200/ComputerCraft/pull/237). They acted a little like Iron Noteblocks from MiscPeripherals (and later MoarPeripherals and Peripherals++), playing any noteblock note.

This in itself is pretty neat, and allows you to recreate [the opening 30 seconds of the only song which matters](https://www.youtube.com/watch?v=FtutLA63Cp8). However, noteblocks only get you so far, and playing the lyrics is impossible. CC:T 1.100 adds support for [playing arbitrary music](https://tweaked.cc/peripheral/speaker.html#v:playAudio) through speakers, allowing us to finally do Bad Apple Right™[^3].

https://user-images.githubusercontent.com/4346137/147082257-5c8709d3-8a20-4d3a-8dcf-d2501c9cd50b.mp4

There's a lot going on under the hood here. The audio is stored as DFPWM, [decoded on the fly to PCM](https://tweaked.cc/library/cc.audio.dfpwm.html) and then given to the speaker[^4]. CC is also just about fast enough to allow you to do some [silly audio processing too](https://twitter.com/CuriousCalamari/status/1467978272684355584), which is neat.

Don't worry if you don't understand any of these words, the documentation contains [a (hopefully) useful introduction](https://tweaked.cc/guide/speaker_audio.html) to all this. Or just ignore all the details, convert your music with [music.madefor.cc](https://music.madefor.cc/) or [LionRay](https://github.com/gamax92/LionRay/), and play it with the built-in `speaker` program!

## Better modem block breaking

When breaking a wired modem and cable, the two parts are destroyed separately. Pre-Minecraft 1.15, the "breaking progress" effect made it clear which part you were breaking, but this was removed as part of the update. Thanks to the power of mixins, we've added this back:

![A ComputerCraft modem being broken](https://user-images.githubusercontent.com/4346137/147380675-0239ffb0-be9e-43e4-b906-8d1454cf58af.png)

## Toodles!

I've been doing this modding lark for a while now. I joined the ComputerCraft forums back in 2014, and started modding only a year later, releasing CCTweaks in early 2015. Plethora and Unborked followed in 2016[^5]. In 2017, ComputerCraft was open sourced (💓 Dan), and CC: Tweaked was first released, which I've been working on ever since (as well as [a](https://emux.cc/) [whole](https://github.com/Squiddev-CC/mbs) [bunch](https://cloud-catcher.squiddev.cc/) [of](https://copy-cat.squiddev.cc/) [other](https://github.com/cc-tweaked/eval.tweaked.cc) [projects](https://music.madefor.cc/)).

This is a lot of words to say that I've been at this for a while, and had a lot of fun in the process, but this is probably it from me as far as CC goes. Modulo any game-breaking bugs and *maybe* new Minecraft versions, this will be the last CC: Tweaked release.

I'm afraid I won't be appointing a successor or whatever. For better or for worse, this has been too much of a "me" project properly let it go. I'm sure some forks will spring up (no pressure [CC-Next](https://github.com/CC-Next/CC-Next)), and of course [CC: Restitched](https://github.com/cc-tweaked/cc-restitched) will continue to thrive.

## In closing

I've not been here for all of it, but it's been a pretty good 10 years for ComputerCraft. As always, thank you to everyone to the community who has contributed to making the mod as special as it is. There's a couple of people I do want to give a special shout out to:

-   Dan and Cloudy, for their work on CC.
-   Lyqyd, BombBloke, and the rest of the OG forum team. It's long dead now, but I met so many talented people (and wonderful friends) on there.
-   @plt-amy, @lemmmy[^6] and the rest of the [SwitchCraft](https://www.technicpack.net/modpack/officialswitchcraft.863735) team for getting CC:T off the ground and keeping me encouraged over the years.

And you, I guess, for getting to the end of this rambling. Hope everyone has a wonderful 2022.

[^1]: GitHub didn't even support footnotes back then. Look how far we've come!

[^2]: Sorry to anyone who was expecting a 2.0.

[^3]: Video is taken from [Axisok's qtccv project](https://github.com/Axisok/qtccv).

[^4]: The audio then re-encoded to DFPWM and sent to the client, but that's more an implementation feature. Does guarantee that you've got the whine and fuzz that Computronics users will be familiar with :).

[^5]: Hah, anyone remember Unborked?! Glad that thing is dead.

[^6]: Lemmmy [has an album out](https://ffm.to/cant-sleep.otw). Go buy it!

**Comment by Lupus590** [comment]
Thanks for the shoutout of CC:N but I have no intention for it to become a place for normal releases. I want it to just be where early features are built. (Although I do recognise how familiar that sounds[^1])

If I knew Java and Forge better and didn't want to do my own MC Mod[^2] then I would offer to step up in maintaining a proper CC:T or a proper fork of it.

[^1]: [CC:T started the same way](http://www.computercraft.info/forums2/index.php?/topic/29161-mc112fork-cc-tweaked/).
[^2]: [Watch this link?](https://github.com/Lupus590-MC/Magic-Computers) But maybe don't get too hopeful, I'm very bad at motivation.

**Reply by SquidDev** [reply]
Oh, I was being serious about by "No pressure". CC:N is the only fork I'm aware of that is trying to be (to any extent) its own thing, but I'm sure others will pop up too. More than happy to add links in the README as they mature. 

Is https://github.com/Lupus590-MC/Magic-Computers private? It 404s for me.

**Reply by Lupus590** [reply]
> Is https://github.com/Lupus590-MC/Magic-Computers private?

It is, I'm not ready for it to be public yet thus "watch this link" and not "see here" or similar.

I've already discussed my plans on MagicComputers (working title, suggestions welcome) on Discord, but that probably was after your seasonal hiatus started.

**Reply by SquidDev** [reply]
Oh, of course. I had parsed that as a literal "there's a video here with some motivation". Did not get enough sleep last night :).

**Comment by Wojbie** [comment]
Allow me to give you heart full thanks for all the time you put (and put up with) into this CC based community and for being a cornerstone that was keeping CC alive for last few years. I wish you good luck in your future endeavors, be they professional or private. Cheers and see ya around! ❤️ Don't be a stranger. 😛


_Sidenote: I love that bad-apple! Speakers are fun!_

**Comment by A-Cloud-Ninja** [comment]
It's sad to see an end to a major chapter of CC, but I'm glad to have around for it. The past 5 or 6 years have been fantastic in the community and I hope it can continue to foster a new generation, even if only in spirit with another mod. Merry Christmas, Happy New Year, Happy Holidays, whatever you celebrate, enjoy everyone :)

**Comment by toad-dev** [comment]
Thank you Squid for all the help with CC: Restitched! I hope ComputerCraft doesn't haunt your dreams for too long, and that whatever you do with your newfound time and mental space brings you joy and contentment. Have a wonderful 2022! 🎉

**Comment by MCJack123** [comment]
Wow, I was not expecting to see this news this morning. Thank you so much Squid for all you've done for us since Dan left - CC wouldn't be alive and thriving today if it wasn't for CC: Tweaked. Good luck in all your future endeavors, and I hope to see you around soon. Have a great new year!

Maybe I should finally get into MC modding after 8 years of waiting...

**Comment by SkyTheCodeMaster** [comment]
Well, I hope you have luck in your future ventures, thank you so much for all you did for us!
1.100 is a very high note for you to leave, with the new speakers.
Have fun!

**Comment by fatboychummy** [comment]
Well, here's going to be another era for me to be nostalgic about in the future!

Squid, you've done an amazing amount of work for this community and I must thank you greatly for continuing ComputerCraft for as long as you have. I wish you good luck in your future endeavors.

**Comment by SirEndii** [comment]
Thank you for the work you did! CC:Tweaked is a nice mod and this new release is amazing.
I'm not good with words, so I'll just say **Thank you** ^^

**Comment by SquidDev** [comment]
Thank you everyone for your kind words! It does mean a lot :).

Just a couple of things I wanted to mention:

 - Please do continue to report bugs! It's _unlikely_ they're going to be fixed by me (unless they're pretty serious, see #1003 and CC:T 1.100.1), but good to have them documented.
 - If anyone has questions about CC:T's internals, don't hesitate to [ding me on IRC/email](https://squiddev.cc/#contact). Or Discord if we're friends I guess? Don't want to leave people entirely in the lurch :).
